include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,1922 @@
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_whatsapp(settings::WhatsappSettings settings, routers::Router *router) noexcept;
+
+    // Event processing methods
+    // ````````````````````````
+
+    /// Blocks until the next event occurs, then routes it.
+    /// \param router Router to handle events. If the router is null, we'll
+    /// use an internal dummy router.
+    void route_next_event(routers::Router *router) noexcept;
+
+    /// Returns true if nettest is done, false otherwise.
+    bool is_done() noexcept;
+
+    /// Interrupts the currently running nettest.
+    void interrupt() noexcept;
+
+  private:
+    // Private methods
+    // ```````````````
+
+    // The default router used when no other router is available.
+    static routers::Router *default_router() noexcept;
+
+    // Internal method called to start any nettest.
+    bool start_common(nlohmann::json, const settings::CommonSettings &, routers::Router *router) noexcept;
+
+    // Private definitions
+    // ```````````````````
+
+    // Deleter for mk_task_t.
+    class TaskDeleter {
+      public:
+        void operator()(mk_task_t *task) noexcept { mk_task_destroy(task); }
+    };
+
+    // Syntactic sugar of a unique mk_task_t pointer.
+    using UniqueTask = std::unique_ptr<mk_task_t, TaskDeleter>;
+
+    // Deleter for mk_event_t.
+    class EventDeleter {
+      public:
+        void operator()(mk_event_t *event) noexcept { mk_event_destroy(event); }
+    };
+
+    // Syntactic sugar for a unique mk_event_t pointer.
+    using UniqueEvent = std::unique_ptr<mk_event_t, EventDeleter>;
+
+    // Private attributes
+    // ``````````````````
+
+    // Mutex used to protect the task_ attribute.
+    std::mutex mutex_;
+
+    // Unique pointer to the FFI task.
+    UniqueTask task_;
+};
+
+/*-
+ * __________        .__               __
+ * \______   \_______|__|__  _______ _/  |_  ____
+ *  |     ___/\_  __ \  \  \/ /\__  \\   __\/ __ \
+ *  |    |     |  | \/  |\   /  / __ \|  | \  ___/
+ *  |____|     |__|  |__| \_/  (____  /__|  \___  >
+ *                                  \/          \/
+ *
+ * Definitions ends here. What follows is the inline implementation of this
+ * API, which you can exclude by setting MK_NETTEST_NO_INLINE_IMPL.
+ *
+ * This is handy if you want to include this code in many translation
+ * units to include the implementation into a single object.
+ */
+#if !defined MK_NETTEST_NO_INLINE_IMPL && !defined SWIG
```

Comment:
I like this pattern of separating interface and implementation keeping all in the same header file - for relatively small files - since it seems to me this simplifies both integration and testing. I'll probably apply this pattern more in the future, to libndt and other small libs we want to integrate.